parser,backupccl: switch RESTORE from kv_option to hardcoded options

adityamaru opened the pull request:
Following in the footsteps of BACKUP, this change switches RESTORE from
using KV option lists to hardcoded options.

The `WITH` clause no longer accepts any set of key/value pairs but only
accepts options allowed by the RESTORE statement.

This change does not introduce any new options but adds the following
enhancements:

1. Ensures that the option is only specified once, otherwise it errors
out.

2. Quoted option names are now disallowed.

Informs: #50735

Release note (backward-incompatible change): Improves WITH option
parsing for RESTORE.  Does not allow the same option to be specified
twice, and also prevents usage of quoted option names.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52302)
<!-- Reviewable:end -->

miretskiy reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
I'm going to add @rohany to take a look as well

> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52302#-:-MDqvhBLBQC30pDxwpXg:b3liivz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt)



<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/parser/sql.y:
```diff
@@ -2314,6 +2329,61 @@ partitioned_backup_list:
+
+// List of valid restore options.
+restore_options:
+  ENCRYPTION_PASSPHRASE '=' string_or_placeholder
```
it seems weird to allow the value to not be a quoted string -- is that intentional?

rohany commented on pkg/sql/parser/sql.y:
```diff
@@ -2314,6 +2329,61 @@ partitioned_backup_list:
+  {
+    $$.val = &tree.RestoreOptions{EncryptionPassphrase: $3.expr()}
+  }
+| INTO_DB '=' database_name
```
do you feel like this is an improvement over the current state of affairs? adding all of these extra keywords when we could just manipulate a map in go seems questionable.

dt replied:
We went this way with BACKUP's option since it makes it more self-documenting -- the parser can actually point to invalid options, the diagram generator can correctly print the valid options, etc

rohany replied:
that makes sense, sounds good.

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @miretskiy, and @rohany)

___
*[pkg/sql/parser/sql.y, line 2362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52302#-MDqLEc4myZjzi1LWEVX-r1-2362:-MDuWoiGCBW_W-v8fxI1:b7t37nl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9cb14aeeb0e69508d41fa615a6d86a5ba6b16bd5/pkg/sql/parser/sql.y#L2362)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

it seems weird to allow the value to not be a quoted string -- is that intentional?
</blockquote></details>

encryption_passphrase can still be specified as a quoted string egs: `encryption_passphrase='abcd'/abcd/"abcd"`, only `into_db` has been switched to accept a non-quoted database name. This is identical to the rule `CREATE DATABASE` accepts.

EDIT: If you were referring to the commit message indicating "Quoted option names are now disallowed." then that was pertaining to the option names not the values. Prior to this change one would have to specify 
`'encryption_passphrase' = 'abcd'`. Now you can only say `encyption_passphrase = 'abcd'`


<!-- Sent from Reviewable.io -->

rohany commented:
I don't think that we should regress on the current behavior though. Many users scripts will have `with into_db = '...'`.

adityamaru commented:
> I don't think that we should regress on the current behavior though. Many users scripts will have `with into_db = '...'`.

Good point I didn't think of that, I'll switch the rule to just point to `string_or_placeholder`, like encryption passphrase.

adityamaru committed:
parser,backupccl: switch RESTORE from kv_option to hardcoded options

Following in the footsteps of BACKUP, this change switches RESTORE from
using KV option lists to hardcoded options.

The WITH clause no longer accepts any set of key/value pairs but only
accepts options allowed by the RESTORE statement.

This change does not introduce any new options but adds the following
enhancements:

Ensures that the option is only specified once, otherwise it errors out.

Quoted option names are now disallowed.

Informs: #50735

Release note (backward-incompatible change): Improves WITH option
parsing for RESTORE. Does not allow the same option to be specified
twice, and also prevents usage of quoted option names.

adityamaru commented:
I'm still working out redaction for the passphrase in the RESTORE job, please hold off on a review. I will ping the PR!

adityamaru commented:
RFAL! @miretskiy some things in restore job description needed modification too so if you could have a quick look too, please 🙂

rohany commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -3762,7 +3762,7 @@ func TestBackupRestoreDropDB(t *testing.T) {
 
 	sqlDB.Exec(t, "BACKUP DATABASE data TO $1", LocalFoo)
 	sqlDB.Exec(t, "CREATE DATABASE data2")
-	sqlDB.Exec(t, "RESTORE data.* FROM $1 WITH OPTIONS ('into_db'='data2')", LocalFoo)
```
We shouldn't regress on any existing behavior -- it should be possible to handle 'into_db' = ... as well.

adityamaru replied:
I was under the impression that one of the expected consequences/improvements of moving away from the stringy map of KVOptions was that the user cannot specify a quoted option name. This would also have been the case after #50723 for all the BACKUP options, which would have previously accepted quotes in the name but now don't as they are hardcoded in the yacc grammar. This is also specified in that PR's description. If this is not the expected behavior then we should probably support all option names to be quoted/unquoted in both BACKUP and RESTORE?

rohany replied:
OK, as long as that was discussed and agreed on

adityamaru replied:
I'll allow @miretskiy to confirm once, I was largely making this change using the BACKUP switch as a reference.

dt replied:
Yeah, we never documented that they were quotable strings -- all the examples in the docs are unquoted, so we said it was okay.

adityamaru commented:
Yevgeniy has intermittent power so @dt can I put this on your plate for a quick look.

miretskiy reviewed (approved):
> Reviewed 10 of 18 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru, @dt, @miretskiy, and @rohany)

___
*[pkg/sql/parser/sql.y, line 2358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52302#-METFsi30hASqN_kJsty:-METFsi30hASqN_kJstz:b-6dwgf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2acea59d2925bc40c128400d26c0d9b8a50f7d/pkg/sql/parser/sql.y#L2358)):*
> ```Yacc
>     $$.val = &tree.Restore{
>       Targets: $2.targetList(), Subdir: $4.expr(), From: $6.listOfStringOrPlaceholderOptList(),
>       AsOf: $7.asOfClause(), Options: *($8.restoreOptions()),
> ```

do you mind formatting like above (one line per option) -- white space is free, but makes it easier to read, imo.
___
*[pkg/sql/sem/tree/backup.go, line 363 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52302#-METGVq3-ujPjHSQYyD5:-METGVq3-ujPjHSQYyD6:bct0kec) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2acea59d2925bc40c128400d26c0d9b8a50f7d/pkg/sql/sem/tree/backup.go#L363)):*
> ```Go
> 		o.SkipMissingFKs == options.SkipMissingFKs &&
> 		o.IntoDB == options.IntoDB &&
> 		o.EncryptionPassphrase == options.EncryptionPassphrase
> ```

why do we need to spell out every field?  Looks like comparisson against RestoreOptions{} should work?


<!-- Sent from Reviewable.io -->

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dt, @miretskiy, and @rohany)

___
*[pkg/sql/parser/sql.y, line 2358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52302#-METFsi30hASqN_kJsty:-METNUGB2tO026fjI-MO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2acea59d2925bc40c128400d26c0d9b8a50f7d/pkg/sql/parser/sql.y#L2358)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do you mind formatting like above (one line per option) -- white space is free, but makes it easier to read, imo.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/backup.go, line 363 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52302#-METGVq3-ujPjHSQYyD5:-METNXeB7Bt3V5lOvtvI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2acea59d2925bc40c128400d26c0d9b8a50f7d/pkg/sql/sem/tree/backup.go#L363)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

why do we need to spell out every field?  Looks like comparisson against RestoreOptions{} should work?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany commented:
lgtm

adityamaru commented:
TFTRs!

bors r+
